### PR TITLE
Tom Barnes' Airport Challenge

### PR DESCRIPTION
**Pull request by forty9er**

### CODE REVIEW

File: lib/plane.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
+require_relative './weather.rb'
+
+class Plane
+
+  include Weather
+
+  attr_reader :location
+  attr_accessor :flying
```
Comment (yevdyko) [inline lib/plane.rb]: Improper use of attr_accessor

File: lib/airport.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+class Airport
+
+  include Weather
+
+  attr_reader :name, :planes, :capacity
+
+  DEFAULT_CAPACITY=10
+
+  def initialize name, capacity=DEFAULT_CAPACITY
```
Comment (danldb) [inline lib/airport.rb]: Look into Ruby's 'keyword arguments' syntax.

File: lib/airport.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+require_relative './weather.rb'
+
+class Airport
+
+  include Weather
```
Comment (danldb) [inline lib/airport.rb]: Use Dependency Injection to add your module, instead of using a mixin.

File: lib/airport.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+  attr_reader :name, :planes, :capacity
```
Comment (danldb) [inline lib/airport.rb]: Do `planes` and `capacity` need to be public methods?

File: lib/airport.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+  def initialize name, capacity=DEFAULT_CAPACITY
+    @name=name
+    @planes=[]
+    @capacity=capacity
+  end
+
+  def full
+    @planes.length>=capacity ? true : false
+  end
```
Comment (danldb) [inline lib/airport.rb]: `#full` should be a private method

File: lib/plane.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+  def initialize
+    @location=nil
+    @flying=true
+  end
+
+  def land airport
+    if @flying==true
+      raise 'Cannot land while airport is full' if airport.full
```
Comment (danldb) [inline lib/plane.rb]: This error depends upon the internal state of the airport - that is knowledge the plane class should not have.

File: lib/plane.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+      raise 'Cannot land while weather is stormy.' if airport.weather==:stormy
```
Comment (danldb) [inline lib/plane.rb]: Again, this guard clasue should exist in the airport, not the plane.

I would recommend doing some research on the idea of 'Encapsulation in OOD'. The basic idea is that objects should know as little as possible about other objects - thus reducing dependency. The overall result of this is modular, reusable and maintainable code.

File: lib/plane.rb [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      complete_landing(airport)
+    else 
+      'This plane is already on the ground.'
```
Comment (danldb) [inline lib/plane.rb]: Don't return a string, raise an error.

File: lib/plane.rb [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    end 
+  end
+
+  def complete_landing airport
+    @location=airport.name
```
Comment (danldb) [inline lib/plane.rb]: There was no need for a location variable, you are going beyond the brief.

File: lib/plane.rb [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    @flying=false
+    airport.planes << self
+    "The plane has landed at #{airport.name}"
```
Comment (danldb) [inline lib/plane.rb]: don't return strings - that would be saved for a user interface, not for an internal object.

File: lib/plane.rb [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+      end
+    else  
+    'The plane is already flying'
+    end  
+  end
+
+  def complete_take_off airport
+    @location=nil
+    @flying=true
+    airport.planes.delete(self)
```
Comment (danldb) [inline lib/plane.rb]: Breaking 'law of demeter'. Also directly modifying internal variables of another object - that is like reaching right in and playing with its privates!

File: lib/weather.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+module Weather
+
+  def weather
+    weather=rand(4)
+    weather==1 ? :stormy : :sunny
+  end
+
+end
```
Comment (danldb) [inline lib/weather.rb]: AWESOME - love the use of a module for this.

File: spec/plane_spec.rb [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+describe Plane do 
+
+  describe 'initialization' do
+    it { is_expected.to respond_to (:location) }
+
+    it { is_expected.to respond_to (:flying) }
+  end
+
+  describe '#take_off' do
+    it { is_expected.to respond_to (:take_off) }
```
Comment (danldb) [inline spec/plane_spec.rb]: respond_to tests are redundent

File: spec/weather_spec.rb [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+  end
+
+  describe '#weather' do
+  
+    it 'is expected to be sunny or stormy' do
+      weather_options=[:sunny, :stormy]
+      expect((weather_options).include?(@dummy.weather)).to be true
+    end 
+  end
+end
```
Comment (danldb) [inline spec/weather_spec.rb]: This is not how I would test the module - but as you are using it as a mixin, then you do need to mix it in.

**Comment by danldb** [comment]
:construction: 

Thanks for the submission Tom. You have really got well developed Ruby skills - your are writing clean, concise code already, at the end of week 1 - well done! There are a bunch of OOD principles that you are violating (you have probably never seen or heard of them before, so don't be too concerned!). Use this week to focus on OOD - that is, SOLID principles and Encapsulation. Things will start to fall into place more once you start to recognise these patterns.

**Pull request closed without merging**